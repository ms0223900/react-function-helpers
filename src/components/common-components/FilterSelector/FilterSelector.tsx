import { 
  Box, 
  Button, 
  makeStyles,
  Typography,
} from '@material-ui/core';
import React from 'react';
import Selects from './SelectsInFilter';
import { FilterSelectorProps } from './types';

const useStyles = makeStyles(theme => ({
  root: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    position: 'relative',
  },
  selectButton: {
    width: '100%',
  },
}));

const FilterSelector = (props: FilterSelectorProps) => {
  const {
    selectedText, 
    isDisplaySelects,
    toggleDisplaySelectsFn,
  } = props;
  const classes = useStyles();

  return (
    <Box className={ classes.root } >
      <Button 
        className={ classes.selectButton }
        variant={ 'outlined' } 
        onClick={toggleDisplaySelectsFn}
      >
        { selectedText }
        <Typography variant={ 'body1' }>
          { ' â–¾ ' }
        </Typography>
      </Button>
      {isDisplaySelects && (
        <Selects 
          {...props}
          closeSelectsFn={toggleDisplaySelectsFn}
        />
      )}
    </Box>
  );
};


export default FilterSelector;
