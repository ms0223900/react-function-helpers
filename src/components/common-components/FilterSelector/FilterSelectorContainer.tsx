import React, {  } from 'react';
import { FilterSelectorContainerProps } from './types';
import FilterSelector from './FilterSelector';
import useFilterSelector from './useFilterSelector';

const FilterSelectorContainer = (props: FilterSelectorContainerProps) => {
  const {
    state,
    handleFilter,
    handleSelect,
    handleToggleDisplaySelects,
  } = useFilterSelector(props);

  return (
    <FilterSelector
      {...state}
      {...props}
      selectFn={handleSelect}
      filterFn={handleFilter}
      toggleDisplaySelectsFn={handleToggleDisplaySelects}
    />
  );
};

export default FilterSelectorContainer;