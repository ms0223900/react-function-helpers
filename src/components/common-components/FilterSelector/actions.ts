import { FilterSelectorContainerStates, FilterSelectorContainerProps } from './types';

export enum FilterSelectorActionTypes {
  SET_STATE,
  TOGGLE_DISPLAY_SELECTS,
  RESET_SELECTED,
  SELECT,
  FILTER,
} 

export interface ToggleDisplaySelectsAction {
  type: FilterSelectorActionTypes.TOGGLE_DISPLAY_SELECTS,
}
export interface SelectAction {
  type: FilterSelectorActionTypes.SELECT,
  payload: SelectPayload
}
export interface ResetSelectAction {
  type: FilterSelectorActionTypes.RESET_SELECTED,
  payload: ResetSelectActionPayload
}
export interface FilterAction {
  type: FilterSelectorActionTypes.FILTER,
  payload: FitlerActionPayload
}
export interface SetStateAction {
  type: FilterSelectorActionTypes.SET_STATE,
  payload: SetStateActionPayload
}
export type FilterSelectorActions = ToggleDisplaySelectsAction | SelectAction | ResetSelectAction | FilterAction | SetStateAction

export interface SelectPayload {
  selectedIndex: number
  options: FilterSelectorContainerProps['options']
  getSelectedOptionFn?: FilterSelectorContainerProps['getSelectedOptionFn']
}
export type ResetSelectActionPayload = FilterSelectorContainerProps['options']
export interface FitlerActionPayload {
  filterInput: FilterSelectorContainerStates['filterInput']
  options: FilterSelectorContainerProps['options']
}
export type SetStateActionPayload = Partial<FilterSelectorContainerStates>;

export const select = (payload: SelectPayload): SelectAction => ({
  type: FilterSelectorActionTypes.SELECT,
  payload,
});

export const resetSelect = (payload: ResetSelectActionPayload): ResetSelectAction => ({
  type: FilterSelectorActionTypes.RESET_SELECTED,
  payload,
});

export const toggleDisplaySelects = (): ToggleDisplaySelectsAction => ({
  type: FilterSelectorActionTypes.TOGGLE_DISPLAY_SELECTS,
});

export const filter = (payload: FitlerActionPayload): FilterAction => ({
  type: FilterSelectorActionTypes.FILTER,
  payload,
});

export const setState = (payload: SetStateActionPayload): SetStateAction => ({
  type: FilterSelectorActionTypes.SET_STATE,
  payload,
});