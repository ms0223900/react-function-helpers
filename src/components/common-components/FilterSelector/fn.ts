import { SelectorOptions } from "./types";

export const defaultTextAndValue = {
  text: '',
  value: '',
  name: ''
};

export const getSelectedRouteNameValue = (
  routes: SelectorOptions, 
  selectedRouteId: string | null | undefined
)  => {
  if(routes) {
    const matchedRoute = routes.find(data => data.value === selectedRouteId);
    if(matchedRoute) {
      return matchedRoute;
    }
  }
  return defaultTextAndValue;
};

export const filterRoutesByValue = (value: string, routes: SelectorOptions) => {
  const filtered = routes.filter(option => option.text.includes(value));
  return value.length === 0 ? routes : filtered;
};