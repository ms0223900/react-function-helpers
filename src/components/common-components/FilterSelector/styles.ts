import { makeStyles } from "@material-ui/core";
import { clickBG } from "../../../styles/styleObjs";


export const useStylesSelectsInFilter = makeStyles(theme => ({
  root: {
    position: 'absolute',
    zIndex: 20202,
    width: '100%',
    // maxWidth: 400,
    backgroundColor: '#fff',
    // textAlign: 'center',
  },

  selects: {
    maxHeight: 400,
    padding: theme.spacing(1),
    overflow: 'auto',
  },

  ...clickBG,
}));