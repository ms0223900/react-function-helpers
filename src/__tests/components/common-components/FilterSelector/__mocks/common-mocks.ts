import { SelectPayload, ResetSelectActionPayload, FitlerActionPayload, SetStateActionPayload } from "../../../../../components/common-components/FilterSelector/actions";
import { SingleSelectorOption, FilterSelectorContainerStates, FilterSelectorContainerProps } from "../../../../../components/common-components/FilterSelector/types";

export const singleSelectorOption: SingleSelectorOption = {
  value: 'a',
  text: 'A',
};

export const selectPayload: SelectPayload = {
  selectedIndex: 0,
  options: [],
};

export const resetSelectActionPayload: ResetSelectActionPayload = [singleSelectorOption];

export const fitlerActionPayload: FitlerActionPayload = {
  filterInput: '',
  options: [],
};

export const setStateActionPayload: SetStateActionPayload = {
  selectedText: 'cjk',
};

export const customInitState: FilterSelectorContainerStates = {
  selectedIndex: 10,
  selectedText: 'abc',
  isDisplaySelects: true,
  filterInput: 'aaa',
  filteredOptions: [
    singleSelectorOption,
    singleSelectorOption,
  ]
};

export const filterSelectorContainerProps: FilterSelectorContainerProps = {
  options: [singleSelectorOption],
  defaultSelectedText: 'bb',
};