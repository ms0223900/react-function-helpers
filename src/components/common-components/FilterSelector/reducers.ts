import { Reducer } from 'all-common-types';
import { FilterSelectorContainerStates } from './types';
import { filterRoutesByValue } from './fn';
import { FilterSelectorActions, FilterSelectorActionTypes } from './actions';
import { initFilterSelectorContainerState } from './useFilterSelector';

type Reducers = Reducer<FilterSelectorContainerStates, FilterSelectorActions, FilterSelectorContainerStates>

const reducer: Reducers = (state, action) => {
  switch (action.type) {
    case FilterSelectorActionTypes.SELECT: {
      const {
        selectedIndex, options, getSelectedOptionFn
      } = action.payload;
      const selected = options[selectedIndex];
      if(selected) {
        getSelectedOptionFn && getSelectedOptionFn(selected);
        return ({
          ...state,
          selectedIndex,
          selectedText: selected.text,
          isDisplaySelects: false,
        });
      }
      return state;
    }

    case FilterSelectorActionTypes.RESET_SELECTED:
      return ({
        ...state,
        filteredOptions: action.payload.options,
        selectedText: action.payload.defaultSelectedText,
        filterInput: initFilterSelectorContainerState.filterInput,
        selectedIndex: initFilterSelectorContainerState.selectedIndex,
      });

    case FilterSelectorActionTypes.FILTER: {
      const { filterInput, options } = action.payload;
      const filteredOptions = filterRoutesByValue(filterInput, options);
      return ({
        ...state,
        filterInput,
        filteredOptions,
      });
    }

    case FilterSelectorActionTypes.TOGGLE_DISPLAY_SELECTS:
      return ({
        ...state,
        isDisplaySelects: !state.isDisplaySelects,
      });

    case FilterSelectorActionTypes.SET_STATE: 
      return ({
        ...state,
        ...action.payload,
      });
  }
};

export default reducer;