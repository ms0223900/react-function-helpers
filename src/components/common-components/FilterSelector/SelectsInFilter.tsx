import React, { memo, useCallback, useRef, createRef, useEffect, forwardRef, Ref, MutableRefObject } from 'react';
import { 
  TextField, 
  List,
  ListItem,
  ListItemText,
  Paper,
  CircularProgress,
  makeStyles,
  Box,
} from '@material-ui/core';
import { SingleSelectProps, SelectsInFilterProps } from './types';
import { useStylesSelectsInFilter } from './styles';
import useScrollToSingleSelect from './useScrollToSingleSelect';

export const NoOptionsInfo = ({ input }: {
  input: string
}) => {
  return (
    <ListItem>
      <ListItemText primary={ `no result matched ${ input }` } />
    </ListItem>);
};

export const SingleSelect = forwardRef(({
  option, isSelected, clickFn
}: SingleSelectProps, ref: Ref<HTMLDivElement>) => {
  return (
    <ListItem  
      ref={ref}
      selected={ isSelected }
      onClick={ clickFn } 
      button
    >
      <ListItemText primary={ option.text } />
    </ListItem>
  );
});

export const Selects = ({ 
  filterInput, 
  filteredOptions, 
  selectedIndex, 
  filterFn, 
  selectFn,
  closeSelectsFn 
}: SelectsInFilterProps) => {
  const listRefs = React.useRef(filteredOptions.map((op) => (
    React.createRef<HTMLDivElement>()
  )));
  const classes = useStylesSelectsInFilter();

  const handleSelectRoute = useCallback((index: number) => {
    return () => {
      selectFn && selectFn(index);
    };
  }, [selectFn]);

  const handleClose = useCallback(() => {
    closeSelectsFn && closeSelectsFn();
  }, [closeSelectsFn]);

  useScrollToSingleSelect(listRefs, selectedIndex);

  return (
    <Box className={ classes && classes.root }>
      <Paper>
        <Box padding={1}>
          <TextField 
            fullWidth 
            autoFocus={ true }
            variant={'outlined'}
            onChange={ filterFn }
            value={ filterInput } />
        </Box>
        <List className={classes.selects}>
          { filteredOptions.length > 0 ? (
            filteredOptions.map((option, i) => (
              <SingleSelect 
                key={i}
                ref={listRefs.current[i]}
                option={option}
                isSelected={selectedIndex === i}
                clickFn={handleSelectRoute(i)}  />
            ))
          ) : (
            <NoOptionsInfo input={filterInput} />
          ) }
        </List>
      </Paper>
      <Box 
        id={'selectsInFilterBG'}
        className={classes.clickBG} 
        onClick={handleClose}
      />
    </Box>
  );
};

export default memo(Selects);