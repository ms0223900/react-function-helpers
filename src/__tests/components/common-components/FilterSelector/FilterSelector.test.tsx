import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer, { act } from 'react-test-renderer';
import FilterSelector from '../../../../components/common-components/FilterSelector/FilterSelector';
import { filterSelectorProps } from './__mocks/common-mocks';
import { Selects } from '../../../../components/common-components/FilterSelector/SelectsInFilter';
import { Button } from '@material-ui/core';

describe('test FilterSelector component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <FilterSelector {...filterSelectorProps} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Selects component should be determined by props.isDisplaySelects', () => {
  it('props.isDisplaySelects is false condition', () => {
    const wrapper = shallow(
      <FilterSelector 
        {...filterSelectorProps}
        isDisplaySelects={false} />
    );
    expect(wrapper.find(Selects)).toHaveLength(0);
  });

  it('props.isDisplaySelects is false condition', () => {
    const wrapper = shallow(
      <FilterSelector 
        {...filterSelectorProps}
        isDisplaySelects={true} />
    );
    expect(wrapper.find(Selects)).toHaveLength(1);
  });
});

describe('test function should be trigger correctly', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  const toggleDisplaySelectsFn = jest.fn();

  it('test toggleDisplaySelectsFn(Button)', () => {
    const wrapper = shallow(
      <FilterSelector 
        {...filterSelectorProps}
        isDisplaySelects={true}
        toggleDisplaySelectsFn={toggleDisplaySelectsFn} />
    );
    expect(toggleDisplaySelectsFn).not.toBeCalled();

    act(() => wrapper.find<any>(Button).props().onClick());
    expect(toggleDisplaySelectsFn).toBeCalled();
  });

  it('test toggleDisplaySelectsFn(Selects)', () => {
    const wrapper = shallow(
      <FilterSelector 
        {...filterSelectorProps}
        isDisplaySelects={true}
        toggleDisplaySelectsFn={toggleDisplaySelectsFn} />
    );
    expect(toggleDisplaySelectsFn).not.toBeCalled();

    act(() => wrapper.find<any>(Selects).props().closeSelectsFn());
    expect(toggleDisplaySelectsFn).toBeCalled();
  });
});

