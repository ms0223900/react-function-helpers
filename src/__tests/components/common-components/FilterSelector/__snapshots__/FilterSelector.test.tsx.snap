// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test FilterSelector component should render correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-4 makeStyles-root-1"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-selectButton-2"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
         â–¾ 
      </p>
    </span>
  </button>
</div>
`;
